pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -3381,19 +3384,40 @@ public void revertReplaceSegments(String tableNameWithType, String segmentLineag
         tableNameWithType, segmentLineageEntryId);
   }
 
-  private void updateSegmentLineageEntryToReverted(String tableNameWithType, SegmentLineage segmentLineage,
-      String segmentLineageEntryId, LineageEntry lineageEntry) {
+  private List<String> deleteOrUpdateSegmentLineageEntryToReverted(String tableNameWithType,
+      SegmentLineage segmentLineage, String segmentLineageEntryId, LineageEntry lineageEntry,
+      List<String> newSegments) {
     // Check that all segments from 'segmentsFrom' are in ONLINE state in the external view.
     Set<String> onlineSegments = getOnlineSegmentsFromExternalView(tableNameWithType);
     Preconditions.checkArgument(onlineSegments.containsAll(lineageEntry.getSegmentsFrom()), String.format(
```

Comment:
We should not have this check when reverting an `IN_PROGRESS` entry. It is possible that the `segmentsFrom` are already expired and removed by the retention manager

Reply:
This should not happen? Currently retention manager will clean up `IN_PROGRESS` lineage entries after 24 hours. Usually the retention period should be > 24 hours.
Say we have segments in `ERROR` state for the segmentsFrom, we probably want to fix the segments first before reverting the entry.

Reply:
I meant the segment retention instead of the lineage cleanup. IIRC we manage segment retention and lineage cleanup separately, and the segments in `segmentsFrom` might already expire.

In high level, when reverting an invalid `IN_PROGRESS` entry, we should not check the `segmentsFrom` but always revert it. For `COMPLETED` entry, we should check to ensure no data loss.

Reply:
Right, we should have the check for `COMPLETED` entries but not `IN_PROGRESS` ones. Removed the check for `deleteOrUpdateSegmentLineageEntryToReverted()` because it's only used for `IN_PROGRESS` entries. Added the if condition to check the lineage entry status in `updateSegmentLineageEntryToReverted()`.